// win32-def/dist/lib/consts/winspool.enum.js
export var PrinterEnumFlags;
(function (PrinterEnumFlags) {
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_DEFAULT"] = 1] = "PRINTER_ENUM_DEFAULT";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_LOCAL"] = 2] = "PRINTER_ENUM_LOCAL";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_CONNECTIONS"] = 4] = "PRINTER_ENUM_CONNECTIONS";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_FAVORITE"] = 4] = "PRINTER_ENUM_FAVORITE";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_NAME"] = 8] = "PRINTER_ENUM_NAME";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_REMOTE"] = 16] = "PRINTER_ENUM_REMOTE";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_SHARED"] = 32] = "PRINTER_ENUM_SHARED";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_NETWORK"] = 64] = "PRINTER_ENUM_NETWORK";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_EXPAND"] = 16384] = "PRINTER_ENUM_EXPAND";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_CONTAINER"] = 32768] = "PRINTER_ENUM_CONTAINER";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICONMASK"] = 16711680] = "PRINTER_ENUM_ICONMASK";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON1"] = 65536] = "PRINTER_ENUM_ICON1";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON2"] = 131072] = "PRINTER_ENUM_ICON2";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON3"] = 262144] = "PRINTER_ENUM_ICON3";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON4"] = 524288] = "PRINTER_ENUM_ICON4";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON5"] = 1048576] = "PRINTER_ENUM_ICON5";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON6"] = 2097152] = "PRINTER_ENUM_ICON6";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON7"] = 4194304] = "PRINTER_ENUM_ICON7";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_ICON8"] = 8388608] = "PRINTER_ENUM_ICON8";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_HIDE"] = 16777216] = "PRINTER_ENUM_HIDE";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_CATEGORY_ALL"] = 33554432] = "PRINTER_ENUM_CATEGORY_ALL";
    PrinterEnumFlags[PrinterEnumFlags["PRINTER_ENUM_CATEGORY_3D"] = 67108864] = "PRINTER_ENUM_CATEGORY_3D";
})(PrinterEnumFlags = PrinterEnumFlags || (PrinterEnumFlags = {}));
